### PR TITLE
HDDS-2285. GetBlock and ReadChunk command from the client should be s…

### PR DESCRIPTION
**Pull request by hanishakoneru**
It can be observed that the GetBlock and ReadChunk command is sent to 2 different datanodes. It should be sent to the same datanode to re-use the connection.

```
19/10/10 00:43:42 INFO scm.XceiverClientGrpc: Send command GetBlock to datanode 172.26.32.224
19/10/10 00:43:42 INFO scm.XceiverClientGrpc: Send command ReadChunk to datanode 172.26.32.231
```

### CODE REVIEW

**Comment by bshashikant** [comment]
Thanks @hanishakoneru for working on this. I am thinking of a case where, let's say two threads trying to read two different keys using the same XceiverClientGrpc object as it is shared object, one thread issues a getBlock for block1 to dn1, dn1 got cached and further read chunk calls for all chunks for all blocks(block1 and block2) will go the same dn. The idea of shuffling the datanodes was basically to distribute the load across the datanodes in a pipeline, but for the same block it should always be tried on the same order of datanodes. I think the patch won't address both these concerns here. What do you think?

**Comment by hanishakoneru** [comment]
Thanks for the review @bshashikant .
I updated the patch to address your comment. I am caching the DN per blockID.

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientGrpc.java [hunk 1]
```diff
@@ -107,6 +111,7 @@ public XceiverClientGrpc(Pipeline pipeline, Configuration config,
         OzoneConfigKeys.OZONE_NETWORK_TOPOLOGY_AWARE_READ_KEY,
         OzoneConfigKeys.OZONE_NETWORK_TOPOLOGY_AWARE_READ_DEFAULT);
     this.caCert = caCert;
+    this.getBlockDNcache = new HashMap<>();
```
Comment (bharatviswa504) [inline hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientGrpc.java]: XceiverClientGrpc is shared across threads, do we need to make this ConcurrentHashMap?

Comment (bshashikant) [reply hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientGrpc.java]: yes, this should be concurrent.

Comment (hanishakoneru) [reply hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientGrpc.java]: Thanks @bharatviswa504.
Fixed it.

**Comment by hanishakoneru** [comment]
Thank you @bshashikant and @bharatviswa504 for the reviews. 
CI run is green. I will merge the PR.

**Pull request merged**